# h12

## a) Kirjoitusvirhe Python-tiedostossa
aloitus 14.00 

postin settings.py tiedostosta import os kohdan ja käynnistin apachen uudelleen

testin tulos:

![image](https://user-images.githubusercontent.com/112497423/222957437-360aae50-2540-4c14-8ca3-c0ac84920597.png)

![image](https://user-images.githubusercontent.com/112497423/222959552-242e4c18-47e8-4edb-a01c-65742fb5ea1c.png)

komennolla ` sudo tail /var/log/apache2/access.log ` tarkistin lokit ja sieltä ilmenee alimalla merkatulla rivillä että importtia "os" ei ole määritetty. ylemmällä rivillä lukee että STATIC ROOT käyttää kyseistä importtia ja sitä ylemmällä rivillä näkyy missä hakemistosta on kyse.

![image](https://user-images.githubusercontent.com/112497423/222960264-e6df57fc-d4c3-4a00-be86-6b21024e8ca5.png)

lisäsin poistetun ` import os ` kohdan takaisin

![image](https://user-images.githubusercontent.com/112497423/222960444-7342994e-3dda-4f74-aefd-726b56d75af3.png)

käynnistin apachen uudelleen komennolla

    sudo systemctl restart apache2
    
testasin toimiiko sivu
    
![image](https://user-images.githubusercontent.com/112497423/222960567-06679c9e-8492-46e4-be1e-7d4d69d1f2ea.png)



## b) Django-projektikansio väärässä paikassa (siis se, jossa on manage.py)
aloitus 14.20

siirsin villeco hakemiston komennolla

![image](https://user-images.githubusercontent.com/112497423/222961680-aff38dfb-36c8-4294-9cb6-bdee5774c558.png)

testasin sivun toimivuutta (ei toimi)

![image](https://user-images.githubusercontent.com/112497423/222961746-7bd3bedf-6444-4b2e-95f0-be1c5f849cb6.png)

virhe loki hain komennolla 

    sudo tail -F /var/log/apache2/access.log
    /var/log/apache2/error.log
    /var/log/apache2/other_vhost.log
    
    
lokista huomaa että hakemistossa /home/ville/publicwsgi/villeco palvelinmääritykset estävät asiakkaan. varmaan sen takia että hakemistoa ei ole enää kyseisellä sijainnilla

![image](https://user-images.githubusercontent.com/112497423/222961846-993607d6-1ee6-4e27-8ab7-9b74eca20855.png)

siirsin hakemiston takaisin alkuperäiselle paikka komennolla
    
![image](https://user-images.githubusercontent.com/112497423/222962095-641dce0b-af17-45f4-8620-c0f4d9ab0cee.png)

sivu toimii suoraan ilman uudelleen käynnistystä

![image](https://user-images.githubusercontent.com/112497423/222962150-14c98da9-3975-435c-bf00-6f80178c6250.png)



## c) Projektikansiolla väärät oikeudet ('chmod ugo-rwx teroco/', 'chmod u+rx teroco/')
aloitus 14.42

aloitin komennolla ` chmod ugo-rwx villeco/ ` joka poistaa kaikilta (u = käyttäjältä g = ryhmältä ja o = muilta kaikki oikeudet (r = luku, w= kirjoitus x= suoritus)

![image](https://user-images.githubusercontent.com/112497423/222963819-ee409506-b99a-42df-880d-62192cf0212d.png)

virhe

![image](https://user-images.githubusercontent.com/112497423/222963879-c0a2b05a-7d38-4a18-ab0f-a411623c239a.png)
 
loki haettu samoilla komennoilla kun aikaisemminkin. ajalta 15.36  tulleista loki ilmoitukset huomaa että pääsy on estetty koska haku luvat puuttuu hakemistoissa /home/ville/publicwsgi/villeco/static ja /home/ville/publicwsgi/villeco/villeco

![image](https://user-images.githubusercontent.com/112497423/222963895-4ced2cf7-396f-49b1-a451-7e27d7f661a5.png)


palautetaan käyttäjän kaikki oikeudet komennolla 

    chmod u+rwx villeco/

palautetaan ryhmän ja muiden luku ja suoritus oikeudut komennolla

    chmod go+rx villeco/
    
oikeudet voi tarkistaa komennolla

    ls -l
    
![image](https://user-images.githubusercontent.com/112497423/222966250-5840586c-d930-4f4c-ba6b-6cfd7932a899.png)    
    
kun oikeudet on palautettu toimii nettisivu taas normaalisti

![image](https://user-images.githubusercontent.com/112497423/222966316-0b124836-da6b-4ec6-a04b-b39ea557794e.png)

## d) Kirjoitusvirhe Apachen asetustiedostossa (/etc/apache2/sites-available/terokarvinen.conf tms)
aloitus 15.40

tein kirjoitus virheen alla näkyvään kohtaan </Directory>

![image](https://user-images.githubusercontent.com/112497423/222967271-d8fafae8-09e9-4397-a3d2-f0969d250585.png)

apachea uudelleen käynnistäessä tulee suoraan viesti mistä huomaa ettei apache toimi oikein. seuraavaksi tein configtestin mikä kertoo suoraan itse aiheutetun kirjoitusvirheen. rivillä 11 olisi pitänyt lukea </Directory> mutta luki </Directoryy>


![image](https://user-images.githubusercontent.com/112497423/222967396-556c0d6c-5e48-4daa-ade1-41381c26ddf4.png)

error loki 

![image](https://user-images.githubusercontent.com/112497423/222967514-3805ef3a-07da-4f31-8a4d-69c9c7cfb51f.png)


tämän kirjoitus virheen aiheuttaa sivun toimimattomuuden

![image](https://user-images.githubusercontent.com/112497423/222967236-9098e3ce-9d45-4fa3-8013-82f8b33eb5f3.png)

seuraavaksi korjasin kirjoitusvirheen, uudelleen käynnistin apachen ja tarkistin sivun toimivuuden.

![image](https://user-images.githubusercontent.com/112497423/222967566-f26a95f6-e1d3-4af4-a7e6-4bbfd15a4a7a.png)


## e) Apachen WSGI-moduli puuttuu ('sudo apt-get purge libapache2-mod-wsgi-py3' tms)

aluksi poistin wsgi modulin alla olevalla komennolla

![image](https://user-images.githubusercontent.com/112497423/222969688-f626027f-d83f-403c-8bd3-2bc96660f344.png)

muutoksen jälkeen sivu ei enää toimi

![image](https://user-images.githubusercontent.com/112497423/222969717-d617209e-094d-41e3-bcba-7adf6d37cef1.png)

lokista ei oikein selvinnyt muuta kuin että apache oli sammunut

![image](https://user-images.githubusercontent.com/112497423/222969802-91ab7760-adaa-4d9e-986c-15606a260b11.png)

configtest kertoo ettei löydä haluttua hakemistoa

![image](https://user-images.githubusercontent.com/112497423/222969851-2840aeb7-f197-4e3a-8624-cbe51aec543b.png)

systemctl statuksesta huomaa että apachen uudelleen käynnistyksen jälkeen syntaxi virhe ja virheellinen komento WSGI

![image](https://user-images.githubusercontent.com/112497423/222969923-14485b7d-d011-4b61-9846-ed70bf0406eb.png)


asensin apachen moduulin uudestaan

-> käynnistin apachen uudestaan 

-> katsoin apachen statuksen uudestaan

![image](https://user-images.githubusercontent.com/112497423/222970015-c272e9ac-dd09-442c-ace2-c33d7928ee4d.png)

kokeilin selaimessa vielä toimivuuden

![image](https://user-images.githubusercontent.com/112497423/222970095-47583d1d-d405-40d1-a3d9-e6f7f7e6dcae.png)


## f) Väärät domain-nimet ALLOWED_HOSTS-kohdassa (settings.py, ja DEBUG=False)

muokkasin settings.py tiedostosta kohtaa ALLOWED HOSTS

![image](https://user-images.githubusercontent.com/112497423/222967973-3636a781-cd90-432d-9564-577c71a2a90f.png)

muutoksen jälkeen 

![image](https://user-images.githubusercontent.com/112497423/222968042-fe07b7cf-eca6-4686-a479-1781d5f1f836.png)


en löytänyt lokista tai muualta mitään mikä olisi täysin selittänyt toimimattomuuden. 

![image](https://user-images.githubusercontent.com/112497423/222968091-3956d671-14b3-4a9a-9018-85d39dc0e2f4.png)

apachen configtestistä ja statuksen perusteella apache toimii. 

![image](https://user-images.githubusercontent.com/112497423/222968687-835134fe-a344-455d-b8d9-c0b42f9c1f4b.png)

korjasin tehtävän alussa tehdyn kirjoitusvirheen

korjauksen jälkeen toimii taas normaalisti 

![image](https://user-images.githubusercontent.com/112497423/222969340-825b4c9b-6ac3-4b8a-bbaa-0f9c8a6057be.png)


## Lähteet

https://terokarvinen.com/
